Implement FastPAM1. Fixes #106

mailology opened the pull request:
I have implemented `FastPAM1` algorithm in this commit. The `FastPAM1` algorithm follows from Algorithm 3 of the PAM and FastPAM1 paper https://arxiv.org/pdf/1810.05691.pdf, where the core idea is to replace the iteration of `k` medoids in the `swap` step with an if conditional outside the loop. This is achievable by using an array to compute `k` loss changes which results in O(N^2) per iteration. 

The algorithm is tested against MNIST-1k dataset as shown below.

<img width="1084" alt="Screen Shot 2021-08-05 at 12 47 17 AM" src="https://user-images.githubusercontent.com/25496074/128376800-40b07204-4e93-407f-90a0-ae527c059a04.png">

It gives the expected set of medoids and the order of `k` is close of zero.

motiwari commented on src/fastpam1.cpp:
```diff
@@ -0,0 +1,193 @@
+/**
```
Include header guards, like [here](https://github.com/ThrunGroup/BanditPAM/blob/b2bfaea933f32ad4ca6acc922e68baf6daa59500/headers/kmedoids_algorithm.hpp#L1-L2)

motiwari commented on headers/kmedoids_algorithm.hpp:
```diff
@@ -185,10 +185,16 @@ class KMedoids {
 
     void fit_naive(arma::mat inputData);
 
+    void fit_fastpam1(arma::mat inputData);
```
Move below `swap_naive`

motiwari commented on setup.py:
```diff
@@ -274,7 +274,8 @@ def build_extensions(self):
          os.path.join('src', 'pam.cpp'),
-         os.path.join('src', 'banditpam.cpp')],
+         os.path.join('src', 'banditpam.cpp'),
+         os.path.join('src', 'fastpam1.cpp')],
```
Add a trailing comma after `)`, then put `],` on a newline (to preserve blame)

motiwari commented on src/fastpam1.cpp:
```diff
@@ -0,0 +1,193 @@
+/**
+ * @file fastpam1.cpp
```
As per [the issue](https://github.com/ThrunGroup/BanditPAM/issues/106), include citations to Schubert's original PAM paper and the FastPAM paper

motiwari commented on src/fastpam1.cpp:
```diff
@@ -0,0 +1,193 @@
+ *
+ * Run the naive FastPAM1 algorithm to identify a dataset's medoids.
+ *
+ * @param input_data Input data to find the medoids of
```
`s/to find the medoids of/to cluster`

motiwari commented on src/fastpam1.cpp:
```diff
@@ -0,0 +1,193 @@
+  data = input_data;
+  data = arma::trans(data);
+  arma::rowvec medoid_indices(n_medoids);
+  // runs build step
```
rm

motiwari commented on src/fastpam1.cpp:
```diff
@@ -0,0 +1,193 @@
+  arma::rowvec medoid_indices(n_medoids);
+  // runs build step
+  KMedoids::build_fastpam1(data, medoid_indices);
+  steps = 0;
```
appears unused; rm?

motiwari replied:
Ah; that function uses a lot of globals. Instead, we should make those member variables. It's ok for this PR and I've updated in #110 for us to fix later.

motiwari commented on src/fastpam1.cpp:
```diff
@@ -0,0 +1,193 @@
+  }
+  medoid_indices_final = medoid_indices;
+  labels = assignments;
+  steps = iter;
```
appears unused; rm?

mailology replied:
We use it to count the number of swap steps and it is used here:
https://github.com/ThrunGroup/BanditPAM/blob/67221ff929d8005e46358a9151795b66bce66891/src/kmedoids_algorithm.cpp#L294-L301

motiwari replied:
OK for this PR; updated #110 to fix

motiwari commented on src/fastpam1.cpp:
```diff
@@ -0,0 +1,193 @@
+  bool medoidChange = true;
+  while (iter < max_iter && medoidChange) {
+    auto previous(medoid_indices);
+    // runs swap step as necessary
```
rm

motiwari commented on src/fastpam1.cpp:
```diff
@@ -0,0 +1,193 @@
+  size_t iter = 0;
+  bool medoidChange = true;
+  while (iter < max_iter && medoidChange) {
+    auto previous(medoid_indices);
```
I think `{brace initialization}` is usually preferred; correct me if I'm wrong

motiwari commented on src/fastpam1.cpp:
```diff
@@ -0,0 +1,193 @@
+    medoidChange = arma::any(medoid_indices != previous);
+    iter++;
+  }
+  medoid_indices_final = medoid_indices;
+  labels = assignments;
```
appears unused; rm?

mailology replied:
`labels` is a global variable which stores the assignment of medoids for each datapoint.
https://github.com/ThrunGroup/BanditPAM/blob/67221ff929d8005e46358a9151795b66bce66891/src/kmedoids_algorithm.cpp#L89-L97

motiwari replied:
OK for this PR; updated #110 to fix

motiwari commented on src/fastpam1.cpp:
```diff
@@ -0,0 +1,193 @@
+ * checks its distance from every other datapoint in the dataset, then checks if
+ * the total cost is less than that of the medoid (if a medoid exists yet).
+ *
+ * @param data Transposed input data to find the medoids of
```
`s/to find the medoids of/to cluster`

motiwari commented on src/fastpam1.cpp:
```diff
@@ -0,0 +1,193 @@
+    double minDistance = std::numeric_limits<double>::infinity();
+    int best = 0;
+    KMedoids::build_sigma(
+           data, best_distances, sigma, batchSize, use_absolute); // computes std dev amongst batch of reference points
```
rm comment

motiwari commented on src/fastpam1.cpp:
```diff
@@ -0,0 +1,193 @@
+        best = i;
+      }
+    }
+    // update the medoid index for that of lowest cost
```
rm

motiwari commented on src/fastpam1.cpp:
```diff
@@ -0,0 +1,193 @@
+/**
+ * \brief Swap step for the FastPAM1 algorithm
+ *
+ * Runs swap step for the FastPAM1 algorithm. Loops over all datapoint and
+ * checks its distance from every other datapoint in the dataset, then checks if
+ * the total cost is less than that of the medoid.
+ *
```
Update comment

motiwari commented on src/fastpam1.cpp:
```diff
@@ -0,0 +1,193 @@
+ * checks its distance from every other datapoint in the dataset, then checks if
+ * the total cost is less than that of the medoid.
+ *
+ * @param data Transposed input data to find the medoids of
```
`s/to find the medoids of/to cluster` throughout

motiwari commented on src/fastpam1.cpp:
```diff
@@ -0,0 +1,193 @@
+  arma::rowvec& medoid_indices,
+  arma::rowvec& assignments)
+{
+  double bestChange = 0; // best loss change 
```
rm comment

motiwari commented on src/fastpam1.cpp:
```diff
@@ -0,0 +1,193 @@
+      delta_td.fill(-di);
+      for (size_t j = 0; j < data.n_cols; j++) {
+          if (j != i) {
+              // compute distance to new medoid
```
rm

motiwari reviewed (changes requested):
Nice work! Definitely a nontrivial change. 

Nits throughout.

mailology commented:
Fixed the nits for [PR#112 ](https://github.com/ThrunGroup/BanditPAM/pull/112). The code is retested and it works as expected.

motiwari reviewed (changes requested):
Suggestions:
- Create `fastpam1.hpp`
- Cite original PAM paper

Everything else OK

mailology commented:
The followings are the changes I have made:

- Cite original PAM paper.
- Introduce `fastpam1.hpp` and create `FastPAM1` subclass inherited from `KMedoids`. There are three member functions in this child: `fit_fastpam1`, `build_fastpam1` and `swap_fastpam1` (they are removed in `KMedoids` class). The definition of these three functions is in `fastpam1.cpp`.
- The usual way we call fit function: `this->*fitFn` seems unnecessarily complicated. In fact, the pointer `this` is from a parent object in the perspective of the fit function `fit_fastpam1` and we need to modify the pointer in order to point to a child function. Also, it turns out that `fitFn` is an intermediate variable which is not needed at all. I have modified the logic so that whenever we do `fit`, the pointer directly points to the corresponding fit function (similar update will be made on `naive` and `BanditPAM` in another PR). 
- I have modified `km::KMedoids::checkAlgorithm` so that it only does checking of the algorithm since we don't need to deal with `fitFn` anymore.

The updated code is checked against Python implementation of `naive`, `BanditPAM` and `FastPAM1` algorithms. All give expected results on MNIST-1k data.

motiwari reviewed (approved):
Looks great, nice work @mailology !
